wpilibNewCommands/src/main/native/cpp/frc2/command/CommandScheduler.cpp

```diff
@@ -196,6 +196,7 @@ void CommandScheduler::Run() {
   m_impl->inRunLoop = true;
   bool isDisabled = frc::RobotState::IsDisabled();
   // Run scheduled commands, remove finished commands.
+  auto toRemove = std::vector<Command*>(m_impl->scheduledCommands.size());
```

Comment:
How often are commands removed?  I would recommend using a SmallVector here with say 8 elements, and not reserving at all.  That will avoid allocating on most calls.

Reply:
> The probable intent is to guarantee the vector has enough space to store all of the commands
That's exactly what I was trying to do.  Thank you for noticing.


> I would recommend using a SmallVector here with say 8 elements

I mean even with a vector, the maximum we'd likely do is allocate at startup, and subsequent reserve calls would invoke malloc, but this wouldn't trigger a system call since we'd already have the space.
and if we were to switch this to Small::Vector

should we do it like this?

```suggestion
  wpi::SmallVector<Command*, 10> toRemove;
```
or like this this?
```suggestion
  wpi::SmallVector<Command*, 10> toRemove;
  toRemove.reserve(m_impl->scheduledCommands.size());
```
(the 10 can be changed to whatever seems appropriate)

Reply:
Default-constructing a `std::vector` and then calling `reserve()` will almost certainly allocate. (I'm not sure what compilers to test, but from https://godbolt.org/z/9sbevjT64, it seems that on x86-64 clang 14.0.0 at least, the starting capacity is 0, so reserving more space will need to allocate more memory)

In contrast, constructing a `SmallVector` and not reserving space (no `reserve()` call) will not allocate any heap memory (instead using stack memory, which is extremely cheap to get) unless the number of commands actually being ended exceeds the template parameter (which is highly unlikely- a typical maximum number of subsystems is 5 to 7, and it's exceedingly rare for all of them to end simultaneously with additional commands also ending simultaneously), in which case it will almost certainly only allocate once. (The number of elements allocated doubles each time the vector grows, so 20 (scheduler-managed) commands would have to end in the same call to `run()` to get just 2 allocations.)

Reply:
but at the same time if a team has decided to go with C++ I am guessing they are going for the highest performance they can get (_as that is mostly the biggest pro of going with C++ in the first place_). which might mean due to our implementation if they have a lot of commands scheduled, it might cause slowdowns that may make the team's devs question why. 

We may also go with robot.py's technique of copying the set and iterating over that instead, which would make erase functions a non-issue but will have the problem of copying a `smallSet` every single periodic loop. Still, if the small set is only stack allocated then it should be fine.

Reply:
I'm not sure I communicated my point clearly enough- What I'm saying is that using a `SmallVector` without reserving will be strictly faster than using a `std::vector` in the vast majority of use cases and at least as fast in the rest, unless someone is misusing the library (in which case we can only help so much).

If you think that more than 16 commands ending all in the same loop is something that we should support, then that's a different question, but my understanding of what you wrote (which might be wrong) is that you were concerned about performance overall without addressing the likelihood of more than 16 commands ending in the same loop. (The 16 comes from 2*8, assuming we use `SmallVector<Command*, 8>`. If you use `SmallVector<Command*, 10>` then the limit becomes 20 commands ending simultaneously before using a small vector gets slower.)